blackbox.cpp

```diff
@@ -322,14 +333,19 @@ int dump(std::ostream &out) {
       out << std::format("{}={}", key, val) << std::endl;
       break;
     }
-    case Type::Invalid:
-      return -EINVAL;
+    case Type::Invalid: {
+      dumped = -EINVAL;
+      goto out;
+    }
     }
 
     dumped++;
     head += hdr->size();
   }
 
+out:
+  lock.unlock();
+  lock_taken = 0;
```

Comment:
Was trying to exercise some C++ muscles, so I had removed some previous use of goto. How about a small class that both acquires the lock and sets the threadlocal in the constructor and then unsets `lock_taken` in destructor? Should make it more C++ ish. Use of goto is annoying in C++ b/c the jump cannot cross any local var initialization.

Reply:
I think in write_locked it's ok as-is b/c it's supposed to be straight line code. Wanted to make it clear that no errors should happen. 

FWIW there's already the InitCleaner thing above for the init stuff. Which is kinda ugly IMO but was just trying stuff out

Reply:
Yep, LGTM. Private constructor would be a good pattern